src/graph/sampler.cc

```diff
@@ -865,4 +882,248 @@ DGL_REGISTER_GLOBAL("sampling._CAPI_LayerSampling")
+      GraphInterface::EdgeArray arr = gptr->FindEdges(worker_seeds);
+      const dgl_id_t *src_ids = static_cast<const dgl_id_t *>(arr.src->data);
+      const dgl_id_t *dst_ids = static_cast<const dgl_id_t *>(arr.dst->data);
+      std::vector<dgl_id_t> src_vec(src_ids, src_ids + num_edges);
+      std::vector<dgl_id_t> dst_vec(dst_ids, dst_ids + num_edges);
+      // TODO(zhengda) what if there are duplicates in the src and dst vectors.
+      nflows[i] = new EdgeBatch();
+
+      if (expand_factor > 0 && num_hops > 0) {
+        nflows[i]->src_nf = new NodeFlow();
+        nflows[i]->dst_nf = new NodeFlow();
+        *nflows[i]->src_nf = SamplerOp::NeighborUniformSample(
+          gptr, src_vec, neigh_type, num_hops, expand_factor, add_self_loop);
+        *nflows[i]->dst_nf = SamplerOp::NeighborUniformSample(
+          gptr, dst_vec, neigh_type, num_hops, expand_factor, add_self_loop);
+      }
+      nflows[i]->pos_subg = new Subgraph();
+      *nflows[i]->pos_subg = gptr->EdgeSubgraph(worker_seeds, false);
+      if (neg_mode.size() > 0) {
+        nflows[i]->neg_subg = new Subgraph();
+        *nflows[i]->neg_subg = NegEdgeSubgraph(gptr->NumVertices(), *nflows[i]->pos_subg,
+                                               neg_mode, neg_sample_size,
+                                               gptr->IsMultigraph(), exclude_positive);
+      }
+    }
+    *rv = WrapVectorReturn(nflows);
+  });
+
+PackedFunc ConvertSubgraphToPackedFunc(const Subgraph& sg);
+
+// Convert EdgeBatch structure to PackedFunc.
+PackedFunc ConvertEdgeBatchToPackedFunc(const EdgeBatch& eb) {
+  auto body = [eb] (DGLArgs args, DGLRetValue* rv) {
+      const int which = args[0];
+      if (which == 0) {
+        *rv = eb.src_nf;
+      } else if (which == 1) {
+        *rv = eb.dst_nf;
+      } else if (which == 2) {
+        *rv = eb.pos_subg;
+      } else if (which == 3) {
+        *rv = eb.neg_subg;
+      } else {
+        LOG(FATAL) << "invalid choice";
+      }
+    };
+  return PackedFunc(body);
+}
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_GetEdgeBatch")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    void *ptr = args[0];
+    const EdgeBatch *eb = static_cast<const EdgeBatch *>(ptr);
+    *rv = ConvertEdgeBatchToPackedFunc(*eb);
+});
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_FreeEdgeBatch")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    void *ptr = args[0];
+    const EdgeBatch *eb = static_cast<const EdgeBatch *>(ptr);
+    delete eb;
+});
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_GetSubgraph")
+ .set_body([] (DGLArgs args, DGLRetValue* rv) {
+     void *ptr = args[0];
+    const Subgraph *eb = static_cast<const Subgraph *>(ptr);
+    *rv = ConvertSubgraphToPackedFunc(*eb);
+ });
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_FreeSubgraph")
+ .set_body([] (DGLArgs args, DGLRetValue* rv) {
+     void *ptr = args[0];
+    const Subgraph *eb = static_cast<const Subgraph *>(ptr);
+    delete eb;
+ });
```

Comment:
Change to object FFI?

Reply:
I'll change it after https://github.com/dmlc/dgl/pull/781 is merged.